rmf_task/include/rmf_task/Task.hpp

```diff
@@ -78,10 +78,39 @@ class Task::Booking
   ///
   /// \param[in] automatic_
   ///   Whether this booking was automatically generated
+  ///
+  /// TODO(AC): Deprecate this constructor
+  Booking(
+    std::string id_,
+    rmf_traffic::Time earliest_start_time_,
+    ConstPriorityPtr priority_,
+    bool automatic_ = false);
+
+  /// Constructor
+  ///
+  /// \param[in] id_
+  ///   The identity of the booking
+  ///
+  /// \param[in] earliest_start_time_
+  ///   The earliest time that the task may begin
+  ///
+  /// \param[in] request_time_
+  ///   The time that this task was booked
+  ///
+  /// \param[in] priority_
+  ///   The priority of the booking
+  ///
+  /// \param[in] requester_
+  ///   The identifier of the entity that requested this task
+  ///
+  /// \param[in] automatic_
+  ///   Whether this booking was automatically generated
   Booking(
     std::string id_,
     rmf_traffic::Time earliest_start_time_,
+    rmf_traffic::Time request_time_,
```

Comment:
I'm not sure if adding a new constructor is the best way forward. If we decide to add more fields in the future, we would need to a constructor each time. Also, these fields are needed for scheduling with rmf-web but it's not apparent from the constructor. Users may still use the older constructor since its available? (Assuming they do not use our dispatcher for some reason). If we always want to ensure these fields are populated, we can consider breaking API and update the existing constructor. This feature will not be backported to `humble` but can be available in  `main` and `iron`.

If we do not want to break API, instead of having two constructors, what do you think about adding `request_time` and `requester` as optional arguments at the end of the arg list? 

```
  Booking(
    std::string id_,
    rmf_traffic::Time earliest_start_time_,
    ConstPriorityPtr priority_,
    bool automatic_ = false,
    std::optional<std::string> requester = std::nullopt,
    std::optional<std::string> request_time = std::nullopt );
  ```
Looking at https://github.com/open-rmf/rmf_api_msgs/pull/35, the new fields are not "required" so we can treat them as such in the API here too.
The dispatcher and fleet adapter can fill/parse these fields if the corresponding field exists in the `task_request.json`

Similarly update the `make()` methods for the various requests.

Reply:
I was hoping to tick-tock the constructor, so next release we remove the old one. I just checked with the project and even though we will be using ROS 2 Humble, we will be using open-rmf `main` source builds, so I'm good with breaking API for `main` and `iron` :+1: 

> Also, these fields are needed for scheduling with rmf-web but it's not apparent from the constructor

Not really, this is just to add more information on `rmf-web`, since the project requires it. So there's no breaking behavior if users don't fill it up

alrighty, I'll get to updating all the PRs soon

Reply:
Actually I was thinking more about it and maybe going with the optional is better cos it will achieve the same end result and we don't need to break API?

Reply:
If we go with the optional with default arguments approach, what would be a good strategy down the road if we want to have these fields as non-optional fields of this class? :thinking: I think request time and requester seem like reasonable fields to be part of booking, and are even critical pieces of information since we log everything and those can be used for tracing activity, unless there were prior plans as to why they were not included in the first place.

for now we will need getters like `std::optional<std::string> requester() const;` but there would be no way to deprecate this getter in a overloading manner over multiple releases since just the return type is changing, it will have to be removed entirely and replaced by `const std::string& requester() const;` when we make it a mandatory field. I'm wondering if it's better to bite the bullet, use the new constructor, mark the old constructor as `[[deprecated]]`, since this is a major release anyway

Reply:
Hmm good point about the getters/setters. I had not considered those.

> I'm wondering if it's better to bite the bullet, use the new constructor, mark the old constructor as [[deprecated]], since this is a major release anyway

One issue I have with the implementation proposed is that if the old constructor is used, we implicitly assign `reuqester` to `""` and `request_time` to `earliest_start_time`.

So I'd prefer adding them as optional args and having the getters/setters remain optional too (semantically makes sense since the system mostly works without these args). Then we update the ros2 side of things such that the json blocks for these fields would not exist in the task_request msg unless concrete values were provided for them.

The one plus side of this is that we can backport to (and release binaries for) `humble`which is gonna be around for the next 4yrs. We can revist breaking API for the next long term release, jazzy, including removing the optional getters/setters for references. 

> unless there were prior plans as to why they were not included in the first place.

Perhaps @mxgrey could advise on this.

Reply:
> we implicitly assign reuqester to "" and request_time to earliest_start_time

yeah, that's not great either haha

Hmm, I guess with breaking API for long term releases on the table, though the optional getters/setters is a little icky, adding optional arguments to the constructor is the ideal solution for now, thanks!

Reply:
👍🏼

Reply:
It appears that while this change would not break API, it might break ABI https://community.kde.org/Policies/Binary_Compatibility_Examples#Change_the_parameters_of_a_function

I guess the safest would be to add another constructor with default args? Atleast that's what is suggested here https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B#The_Do's_and_Don'ts (search "Extending a function with another parameter")

Reply:
I was hoping to avoid all this by squeeze this in before a release so we don't have to worry about breaking ABI. But oh well. I'll make the changes with overloads

Reply:
@Yadunund  https://github.com/open-rmf/rmf_task/pull/81/commits/1b0b05bb02973389fe8ebe6e9d98fb910d183d8a

Reply:
I would say the `iron` release is new enough that we can still break ABI if needed. So i'll be okay with backporting any API/ABI breaking changes into `iron`.

Reply:
Just noticed that all the requests are in the same situation too, that means overloading for `ChargeBattery::make`, `Clean::make`, `Delivery::make`, `Loop::make`, etc is required too. I'll need to overload them as well (similar to the original implementation before any reviews), for sanity check lmk if that's the way to go before I make the changes